[WIP] Add support for validating unlisted add-ons as gatekeeper for auto...

kmaglione opened the pull request:
...mated signing.

Submitting early WIP with some of the required tests for feedback/testing Olympia code.

(CC @magopian)

magopian commented on tests/resources/installrdf/shouldnt_exist.rdf:
```diff
@@ -8,6 +8,7 @@
 	<em:type>2</em:type>
 	
 	<em:updateURL>true</em:updateURL>
+	<em:updateKey>meh</em:updateKey>
```
Didn't know about this one ;) :+1:

kmaglione replied:
We just never checked for it before. I added it because we're now warning when it's missing, for unhosted add-ons.

magopian commented on validator/errorbundler.py:
```diff
@@ -74,6 +72,8 @@ def __init__(self, determined=True, listed=True, instant=False,
 
         self.supported_versions = self.for_appversions = for_appversions
 
+    SEVERITIES = {"trivial", "low", "medium", "high"}
```
I don't remember seeing anything about `trivial` in your notes: I remember that a low severity would still let an addon be automatically reviewed (pass validation) if it had the exact same warning in a previously reviewed version. How about `trivial`?

kmaglione replied:
I haven't decided if I'm going to keep that yet. I added it for the moment because there was a warning that was signing-related, but I didn't want to trigger manual review. If it stays, the idea would be that it would show up as a signing-related message, but wouldn't prevent signing. Or maybe a bunch of them would, but a single one wouldn't.

magopian commented on validator/errorbundler.py:
```diff
@@ -439,7 +447,7 @@ def _print_message(self, prefix, message, verbose=True):
                 verbose_output.append("\tColumn:\t%d" % message["column"])
 
-            if "context" in message and message["context"]:
+            if message.get("context"):
```
:+1:

magopian commented on validator/testcases/installrdf.py:
```diff
@@ -53,7 +53,6 @@ def _test_rdf(err, install):
                       "updateInfoURL",
                       "updateKey",
                       "updateURL",
-                      "updateLink",  # Banned, but if not, pass it once.
```
Is that deprecated?

kmaglione replied:
It's never actually existed. It's an element in update.rdf, but not in install.rdf. Donno why it was there.

magopian commented on validator/testcases/installrdf.py:
```diff
@@ -82,16 +81,19 @@ def _test_rdf(err, install):
     for pred_raw in install.rdf.predicates(top_id, None):
         predicate = pred_raw.split("#")[-1]
 
+        value = install.get_object(top_id, pred_raw)
```
nice factorization :+1:

magopian commented on validator/testcases/javascript/actions.py:
```diff
@@ -81,16 +78,26 @@ def trace_member(traverser, node, instantiate=False):
+                    kw.update(CONTRACT_ENTITIES[identifier])
+
+                    traverser.warning(**kw)
+
```
What does this change do? Is it related to the bug?

magopian replied:
Ah, I see it's another test that you added a severity warning too, below.

magopian commented on validator/testcases/javascript/predefinedentities.py:
```diff
@@ -579,13 +579,30 @@
-                      }}}
+                      }
+
+CONTRACT_ENTITIES = {}
```
This feels a bit weird, why not remove this line, and on line 600 do something like
`CONTRACT_ENTITIES = dict(....)`

kmaglione replied:
I went back and forth on that one. I mainly just wanted the `CONTRACT_ENTITIES` and `INTERFACE_ENTITIES` lines to follow the same pattern. But I'm happy to change it to just a `dict()` on that line.

magopian commented:
This looks promising! Do you plan on having a high level counter for the different severities, in the json returned? Similar to what's already done for errors/warnings, where it says how many there are without digging/looping on the messages themselves?

kmaglione commented:
I hadn't planned on it, but it sounds like a good idea.

kmaglione commented:
Updated WIP. Addressed most comments and added some more checks. I made some non-trivial changes to the way the `"readonly"` property works (the old behavior had a lot of problems), so that might be worth giving some extra attention.

magopian commented on tests/helper.py:
```diff
@@ -90,18 +90,29 @@ def assert_failed(self, with_errors=False, with_warnings=None):
+        def test_messages(mtype, expected):
+            messages = getattr(self.err, mtype)
+
+            if isinstance(expected, (list, tuple)):
```
This could be good enough for test code, but the proper way is to do something like http://stackoverflow.com/a/1952655

kmaglione replied:
On May 5, 2015 6:47 AM, "Mathieu Agopian" <notifications@github.com> wrote:
>
> In tests/helper.py:
>
> > -                assert self.err.warnings, "Warnings were expected."
> > -            elif self.err.warnings:
> > -                raise ("Tests found unexpected warnings: %s" %
> > -                           self.err.print_summary())
> > +        def check_message(message, props):
> > +            return all(message.get(prop) == value
> > +                       for prop, value in props.iteritems())
> > +
> > +        def find_message(messages, props):
> > +            return any(check_message(message, props)
> > +                       for message in messages)
> > +
> > +        def test_messages(mtype, expected):
> > +            messages = getattr(self.err, mtype)
> > +
> > +            if isinstance(expected, (list, tuple)):
>
> This could be good enough for test code, but the proper way is to do
something like http://stackoverflow.com/a/1952655

I'd already changed this locally :)

magopian commented on tests/helper.py:
```diff
@@ -90,18 +90,29 @@ def assert_failed(self, with_errors=False, with_warnings=None):
+            if isinstance(expected, (list, tuple)):
+                assert all(find_message(messages, props)
+                           for props in expected)
+            elif expected:
```
Would that read better if it was written `elif expected is True:`?

kmaglione replied:
On Tue, May 05, 2015 at 06:48:58AM -0700, Mathieu Agopian wrote:
>Would that read better if it was written `elif expected is True:`?

Maybe, but it generally feels like a bad practice to treat 
truthy values that aren't `True` as false.

magopian commented on tests/helper.py:
```diff
@@ -90,18 +90,29 @@ def assert_failed(self, with_errors=False, with_warnings=None):
+                           for props in expected)
+            elif expected:
+                assert messages, "Expected %s." % mtype
+            elif expected is not None:
```
That would benefit from a comment like `if expected is False, or falsy like [], {}...`

magopian replied:
Correction: if it's an iterable, it'll be treated by line 104. So the comment could be `if expected is False or 0`. It could also be rewritten `if expected in [False, 0]`?

Also, why not accept `None`? That would allow calling `test_messages('errors', None)` which is pretty clear? In that case the code would simply read `else:  # If expected is falsy.`.

kmaglione replied:
On Tue, May 05, 2015 at 07:05:49AM -0700, Mathieu Agopian wrote:
>> +            return all(message.get(prop) == value
>> +                       for prop, value in props.iteritems())
>> +
>> +        def find_message(messages, props):
>> +            return any(check_message(message, props)
>> +                       for message in messages)
>> +
>> +        def test_messages(mtype, expected):
>> +            messages = getattr(self.err, mtype)
>> +
>> +            if isinstance(expected, (list, tuple)):
>> +                assert all(find_message(messages, props)
>> +                           for props in expected)
>> +            elif expected:
>> +                assert messages, "Expected %s." % mtype
>> +            elif expected is not None:
>
>Correction: if it's an iterable, it'll be treated by line 104. So the comment could be `if expected is False or 0`. It could also be rewritten `if expected in [False, 0]`?
>
>Also, why not accept `None`? That would allow calling `test_messages('errors', None)` which is pretty clear? In that case the code would simply read `else:  # If expected is falsy.`.

See the docstring. It took me a while to decipher the behavior 
of the old version, even with its docstring, but essentially 
`None` means we don't care whether there are warnings/errors, 
while `False` means we explicitly expect there to be none.

kmaglione replied:
On Tue, May 05, 2015 at 07:05:49AM -0700, Mathieu Agopian wrote:
>Also, why not accept `None`? That would allow calling `test_messages('errors', None)` which is pretty clear? In that case the code would simply read `else:  # If expected is falsy.`.

Ah, sorry, I now see that I finished updating that docstring 
after I last pushed. I thought I'd done it before. For 
reference, then:

```python
        """
        Asserts that the error bundle has registered a failure. If
        `with_warnings` is any value any true value, or `None`, a warning
        is considered a failure.

        `with_warnings` or `with_errors` may be any of the following:

        * True: Messages of this type must be present.
        * False: Messages of this type must not be present.
        * None: Messages of this type may or may not be present.
        * Iterable of dicts: For dict returned by the iterator, at least
        one message must have a matching item for every key/value pair in the
        dict.
        """
```

(Note that this behavior, with relation to `True`/`False`/`None` 
is unchanged. I just made the behavior consistent between 
warnings and errors, and allowed specifying specific messages 
which were expected.)

magopian replied:
Where is this docstring? There's a typo (`is any value any true value`)

kmaglione replied:
It replaces the docstring for `assert_failed`, but like I said, I added it just after I pushed. It will be in my next update later today.

magopian commented on tests/helper.py:
```diff
@@ -90,18 +90,29 @@ def assert_failed(self, with_errors=False, with_warnings=None):
-            elif self.err.warnings:
-                raise ("Tests found unexpected warnings: %s" %
-                           self.err.print_summary())
+        def check_message(message, props):
```
Docstring or comment please

magopian commented on tests/helper.py:
```diff
@@ -90,18 +90,29 @@ def assert_failed(self, with_errors=False, with_warnings=None):
+            return all(message.get(prop) == value
+                       for prop, value in props.iteritems())
+
+        def find_message(messages, props):
```
Docstring or comment please

magopian commented on tests/test_js_functions.py:
```diff
@@ -57,6 +57,38 @@ def test_createElement():
+
+    eq_(err.warnings[0]["id"],
+        ("js", "traverser", "dangerous_global"))
+    eq_(err.warnings[0].get("signing_severity"), "high")
```
why a `.get()` here?

kmaglione replied:
On Tue, May 05, 2015 at 07:07:48AM -0700, Mathieu Agopian wrote:
>> @@ -57,6 +57,38 @@ def test_createElement():
>>      document.createElement();
>>      """)
>>
>> +def test_enablePrivilege():
>> +    """Tests that enablePrivilege throws a warning."""
>> +
>> +    err = _do_test_raw("""
>> +        netscape.security.PrivilegeManager.enablePrivilege()
>> +    """)
>> +    assert err.warnings
>> +
>> +    eq_(err.warnings[0]["id"],
>> +        ("js", "traverser", "dangerous_global"))
>> +    eq_(err.warnings[0].get("signing_severity"), "high")
>
>why a `.get()` here?

Not all messages have a `"signing_severity"` property. I guess 
the `KeyError` would make the test fail the same as the 
`AssertionError`, I just tend to avoid letting tests fail that 
way.

magopian replied:
Makes sense

magopian commented on tests/test_js_functions.py:
```diff
@@ -57,6 +57,38 @@ def test_createElement():
+
+        eq_(err.warnings[0]["id"],
+            ("js", "traverser", "dangerous_global"))
+        eq_(err.warnings[0].get("signing_severity"), "low")
```
Same here, why a `.get()`?

kmaglione replied:
On Tue, May 05, 2015 at 07:10:11AM -0700, Mathieu Agopian wrote:
>Did you mean to make it a `low` signing severity?

Yes. I think this is relatively benign, but I could be convinced 
otherwise.

magopian commented on tests/test_js_functions.py:
```diff
@@ -57,6 +57,38 @@ def test_createElement():
+        ("js", "traverser", "dangerous_global"))
+    eq_(err.warnings[0].get("signing_severity"), "high")
+
+def test_privileged_unprivileged_interacion():
```
typo `interaction`

magopian commented on tests/test_js_signing.py:
```diff
@@ -0,0 +1,55 @@
+Various tests related to validation for automated signing.
+"""
+
+from js_helper import TestCase
```
`from .js_helper import TestCase`

kmaglione replied:
On Tue, May 05, 2015 at 07:15:07AM -0700, Mathieu Agopian wrote:
>> @@ -0,0 +1,55 @@
>> +"""
>> +Various tests related to validation for automated signing.
>> +"""
>> +
>> +from js_helper import TestCase
>
>`from .js_helper import TestCase`

I copied this from a different testcase. Validator hasn't used 
new-style relative imports in the past, but I suppose now that 
we're on 2.7 everywhere...

magopian replied:
Yeah, if we have opportunities to improve the code we're copying/modifying anyways... let's take them ;)

magopian commented on tests/test_js_signing.py:
```diff
@@ -0,0 +1,55 @@
+                "Services.search")
+
+        for obj in objs:
+            warnings = {"signing_severity": "high",
```
This can be declared out of the loop, one line above

kmaglione replied:
On Tue, May 05, 2015 at 07:20:43AM -0700, Mathieu Agopian wrote:
>This can be declared out of the loop, one line above

One of the members gets modified a few lines down. If I moved it 
out of the for loop, I'd have to revert those changes at the 
start of each iteration.

magopian commented on tests/test_js_signing.py:
```diff
@@ -0,0 +1,55 @@
+                               "changes")},
+
+            for prop in "currentEngine", "defaultEngine":
+                yield test, obj, prop, " = foo", warnings
```
What does that do? It looks like some kind of hackery to "generate" tests for unittest/nose? I don't like it, I'm sure you can do the same with a simple call to your `test()` helper above.

To be honest, I'd rather have two separate tests here, and no loop, no helper.

kmaglione replied:
On Tue, May 05, 2015 at 07:24:06AM -0700, Mathieu Agopian wrote:
>What does that do? It looks like some kind of hackery to "generate" tests for unittest/nose? I don't like it, I'm sure you can do the same with a simple call to your `test()` helper above.
>
>To be honest, I'd rather have two separate tests here, and no loop, no helper.

I'm just following the same pattern as the more-recently-added 
tests. I think the idea is that multiple values can be tested 
without one failure short-circuiting the rest.

I don't really have a strong opinion, though. I was just trying 
to be consistent with newer code.

magopian commented on tests/test_js_signing.py:
```diff
@@ -0,0 +1,55 @@
+
+        for method in ("create", "createChild", "writeBinaryValue",
+                       "writeInt64Value", "writeIntValue", "writeStringValue"):
+            yield test, method
```
you could get rid of the `yield` hack and paste the content of the `test()` helper in the loop.

kmaglione replied:
On Tue, May 05, 2015 at 07:31:25AM -0700, Mathieu Agopian wrote:
>you could get rid of the `yield` hack and paste the content of the `test()` helper in the loop.

See above. The `yield` hack causes tests to continue even after 
one fails. I'm not even sure how much of a hack it is. It's a 
feature of nose.

magopian replied:
We don't care if it fails straight away or continue, right? If the suite fails, we still have to fix, so I'd rather not have code like that.

kmaglione replied:
Yes, but if it fails and short-circuits the rest of the tests, it's not always clear which parts are failing, which often means multiple runs after partial fixes to see what's going on. This is particularly true for the other test cases that check both `Services.*` and `.getService(*)`, where the failure could be in the definition on the `Services` object or in the definition of the interface. If all of the tests run regardless of one failure, it's much more obvious where the problem lies before you start digging into code.

magopian commented on tests/test_js_xpcom.py:
```diff
@@ -221,32 +221,6 @@ def test_xpcom_shortcut_services_ww():
-    foo.defaultEngine = 123;
-    """).failed()
-
-
```
Is that linked to this PR? If not, please split those in a separate PR.

kmaglione replied:
On Tue, May 05, 2015 at 07:32:32AM -0700, Mathieu Agopian wrote:
>Is that linked to this PR? If not, please split those in a separate PR.

Yes. This tests exactly the opposite of what it's supposed to be 
testing, so one of my new checks causes it to fail.

magopian commented on validator/testcases/javascript/actions.py:
```diff
@@ -81,16 +78,26 @@ def trace_member(traverser, node, instantiate=False):
             if "xpcom_wildcard" in base.value:
                 traverser._debug("MEMBER_EXP>>XPCOM_WILDCARD")
+
+                from predefinedentities import CONTRACT_ENTITIES
```
Why not put that at the top of the module? Import loop? If so, please comment

kmaglione replied:
On Tue, May 05, 2015 at 07:42:16AM -0700, Mathieu Agopian wrote:
>>          # Handle the various global entity properties.
>>          if base.is_global:
>>              # If we've got an XPCOM wildcard, return a copy of the entity.
>>              if "xpcom_wildcard" in base.value:
>>                  traverser._debug("MEMBER_EXP>>XPCOM_WILDCARD")
>> +
>> +                from predefinedentities import CONTRACT_ENTITIES
>
>Why not put that at the top of the module? Import loop? If so, please comment

I'm not sure. There are 6 other local imports of 
`predefinedentities` in this file, and I didn't have the time to 
try to find any side-effects of making them global when I was 
writing this, so I opted to stay on the safe side.

magopian commented on validator/testcases/javascript/actions.py:
```diff
@@ -81,16 +78,26 @@ def trace_member(traverser, node, instantiate=False):
+                from predefinedentities import CONTRACT_ENTITIES
+                if identifier in CONTRACT_ENTITIES:
+                    kw = dict(err_id=("js", "actions", "dangerous_contract"),
+                              warning="Dangerous XPCOM contract ID")
```
Why not define all the dict items in the same place (either here or in `prefedinedentities`)?

kmaglione replied:
On Tue, May 05, 2015 at 07:46:49AM -0700, Mathieu Agopian wrote:
>>          # Handle the various global entity properties.
>>          if base.is_global:
>>              # If we've got an XPCOM wildcard, return a copy of the entity.
>>              if "xpcom_wildcard" in base.value:
>>                  traverser._debug("MEMBER_EXP>>XPCOM_WILDCARD")
>> +
>> +                from predefinedentities import CONTRACT_ENTITIES
>> +                if identifier in CONTRACT_ENTITIES:
>> +                    kw = dict(err_id=("js", "actions", "dangerous_contract"),
>> +                              warning="Dangerous XPCOM contract ID")
>
>Why not define all the dict items in the same place (either here or in `prefedinedentities`)?

Well, the main reason is that I don't want to have to have to 
duplicate keys.

At the moment, all of the contract IDs in that dict have the 
same properties, so that's not an issue. However, it's always 
worried me that we tend not to do checks for contract IDs, only 
interface names, so I intend to add more in the future.

The slightly more complicated reason is that most of the 
definitions in `predefinedentities` follow a similar pattern, 
where they can accept a string, for only the description, a 
dict, which can define other properties as well, or a callable 
which... has completely different behavior depending on exactly 
where you use it.

All of this is extremely messy and inconsistent, so what I'd 
really like to do is unify the handling of all of them via some 
kind of helper function, which means any new definitions along 
these lines need to follow the same behavior in any case.

magopian commented on validator/testcases/javascript/actions.py:
```diff
@@ -595,7 +602,9 @@ def _expr_assignment(traverser, node):
         global_overwrite = False
-        readonly_value = True
+        # FIXME(Kris): This is not a great heuristic.
+        from predefinedentities import is_shared_scope
```
Same question as above, why not at the top of the module?

magopian commented on validator/testcases/javascript/predefinedentities.py:
```diff
@@ -64,6 +101,12 @@
 ]
 
 
+def is_shared_scope(t, r=None, rn=None):
```
Docstring and no one-letter variables please

mstriemer replied:
I don't know what a "shared scope" is. Can you please explain that in the docstring?

kmaglione replied:
On Tue, May 05, 2015 at 07:49:35AM -0700, Mathieu Agopian wrote:
>> +def is_shared_scope(t, r=None, rn=None):
>
>Docstring and no one-letter variables please

I'd like to, but unfortunately these are passed as keyword args, 
so changing them requires massive changes elsewhere.

-- 
Kris Maglione

The trouble with the world is that the stupid are cocksure and the
intelligent full of doubt.
	--Bertrand Russell

kmaglione replied:
On Tue, May 05, 2015 at 07:57:46AM -0700, Mark Striemer wrote:
>I don't know what a "shared scope" is. Can you please explain that in the docstring?

I'll add a docstring, but essentially:

Most of these properties are flagged because when add-ons load 
code into a shared scope, such as a window, or web page, and 
replace or modify them, they cause conflicts with other code. 
This heuristic is the same as the previous version of the code, 
and assumes (incorrectly) that bootstrapped add-ons will never 
load code into a shared scope, and otherwise only recognizes 
JavaScript modules as private scopes.

The main change in this code is that rather than assume that 
`"readonly"` in general only applies to shared scopes, it now 
needs to be explicitly specified.

magopian replied:
Is that clearer for you @mstriemer ? I'm still feeling confused about all this, but I don't have any context, and don't know the amo-validator code.

mstriemer replied:
This seems like it is moving the conditional from `is_shared_scope() and readonly_value` to when `readonly_value` is set which probably makes sense. I don't see `readonly` being set to anything other than this though but maybe I'm missing something?

kmaglione replied:
It's set to a constant value in a lot of places. The only other place it's given a function is:

```py
    u"top": {"readonly": actions._readonly_top},
```

Though that case just emits a notice.

But that's only two instances, so I guess it shouldn't be a problem to change the argument names after all. I was thinking of `"dangerous"`, where lambdas are used much more often.

magopian commented on validator/testcases/javascript/actions.py:
```diff
@@ -595,7 +602,9 @@ def _expr_assignment(traverser, node):
-        readonly_value = True
+        # FIXME(Kris): This is not a great heuristic.
+        from predefinedentities import is_shared_scope
+        readonly_value = is_shared_scope(traverser)
```
This is the non-trivial change to the `readonly` part you were talking about? @mstriemer could you please have a look?

magopian commented on validator/testcases/javascript/actions.py:
```diff
@@ -643,33 +651,36 @@ def _expr_assignment(traverser, node):
-            not traverser.err.get_resource("em:bootstrap") and
-            readonly_value == True):
+        if callable(readonly_value):
+            # The readonly attribute supports a lambda function that accepts
```
It's missing the end of the sentence?

kmaglione replied:
On Tue, May 05, 2015 at 07:58:18AM -0700, Mathieu Agopian wrote:
>>
>> -        if (global_overwrite and
>> -            not traverser.is_jsm and
>> -            not traverser.err.get_resource("em:bootstrap") and
>> -            readonly_value == True):
>> +        if callable(readonly_value):
>> +            # The readonly attribute supports a lambda function that accepts
>
>It's missing the end of the sentence?

That it is. Unfortunately, I'm not entirely sure what it's 
supposed to be. I didn't actually change this line, just the 
code around it.

magopian commented on validator/testcases/javascript/actions.py:
```diff
@@ -643,33 +651,36 @@ def _expr_assignment(traverser, node):
-                filename=traverser.filename,
-                line=traverser.line,
-                column=traverser.position,
-                context=traverser.context)
```
We don't need those anymore?

kmaglione replied:
On Tue, May 05, 2015 at 08:00:20AM -0700, Mathieu Agopian wrote:
>>                  err_id=("testcases_javascript_actions",
>>                          "_expr_assignment",
>>                          "global_overwrite"),
>>                  warning="Global variable overwrite",
>>                  description="An attempt was made to overwrite a global "
>> -                            "variable in some JavaScript code.",
>> -                filename=traverser.filename,
>> -                line=traverser.line,
>> -                column=traverser.position,
>> -                context=traverser.context)
>
>We don't need those anymore?

No. mstriemer added these utility functions to `Traverser` a 
while ago to take care of those parameters, and has been slowly 
fixing legacy calls since.

magopian commented on validator/testcases/javascript/actions.py:
```diff
@@ -939,7 +949,7 @@ def _get_as_num(value):
         else:
             return int(value)
-    except ValueError, TypeError:
+    except (ValueError, TypeError):
```
:+1:

kmaglione replied:
On Tue, May 05, 2015 at 08:01:30AM -0700, Mathieu Agopian wrote:
>> -    except ValueError, TypeError:
>> +    except (ValueError, TypeError):
>
>:+1:

One of the benefits of skimming through the copious output of 
`flake8`...

magopian commented on validator/testcases/javascript/predefinedentities.py:
```diff
@@ -11,52 +12,88 @@
-                         "cause of deadlocks, crashes, and other errors due to "
-                         "unintended reentrancy. Please use asynchronous callbacks "
-                         "instead wherever possible",
+    u"newThread": (
```
You don't need parens here to continue the strings.

kmaglione replied:
On Tue, May 05, 2015 at 08:04:47AM -0700, Mathieu Agopian wrote:
>You don't need parens here to continue the strings.

I've added parens to most of these descriptions in the past, 
since they can optionally be tuples of strings rather than 
single strings, and it seemed sensible to use parens for all of 
them. I guess I don't really have an objection to removing them, 
though.

magopian commented on validator/testcases/javascript/predefinedentities.py:
```diff
@@ -11,52 +12,88 @@
+        "Extensions must not alter user preferences such as the current home "
+        "page, new tab page, or search engine, without explicit user consent, "
+        "in which a user takes a non-default action. Such changes must also "
+        "be reverted when the extension is disabled or uninstalled.",
```
Did you mean to have a tuple here?

kmaglione replied:
On Tue, May 05, 2015 at 08:07:15AM -0700, Mathieu Agopian wrote:
>Did you mean to have a tuple here?

Yes. Tuple descriptions are rendered as one paragraph per 
element.

magopian commented on validator/testcases/javascript/predefinedentities.py:
```diff
@@ -11,52 +12,88 @@
+    (u"network.http.", NETWORK_PREF_MESSAGE),
+    (u"network.websocket.", NETWORK_PREF_MESSAGE),
     (u"nglayout.debug.disable_xul_cache", None),
+    # Let's see if we can get away with this...
```
What do you mean?

kmaglione replied:
On Tue, May 05, 2015 at 08:09:14AM -0700, Mathieu Agopian wrote:
>>      (u"nglayout.debug.disable_xul_cache", None),
>> +    # Let's see if we can get away with this...
>
>What do you mean?

I'm slightly worried that this substring may come up in 
too many unrelated contexts, but I'd prefer not to make it more 
specific, since basically any preference in that branch should 
be flagged.

magopian commented on validator/testcases/javascript/predefinedentities.py:
```diff
@@ -88,7 +131,59 @@
+
+def search_warning(severity="medium", editors_only=False,
+                   message=SEARCH_MESSAGE,
+                   description=SEARCH_DESCRIPTION):
```
You aren't using any other parameters than `severity` in your calls, do we still need those extra ones here? Are you planning on using them in this PR?

I'd advise against premature optimization.

kmaglione replied:
On Tue, May 05, 2015 at 08:14:14AM -0700, Mathieu Agopian wrote:
>You aren't using any other parameters than `severity` in your calls, do we still need those extra ones here? Are you planning on using them in this PR?

Initially I was planning on using more of them, yes, and I still 
may before I'm done. I'll simplify if it turns out they're not 
needed.

magopian commented on validator/testcases/javascript/predefinedentities.py:
```diff
@@ -575,32 +689,72 @@
 
 def build_quick_xpcom(method, interface, traverser, wrapper=False):
     """A shortcut to quickly build XPCOM objects on the fly."""
+    extra = ()
```
Seems like it got quite a few changes, can you please document what it's doing now? Constructing "deep" objects?

magopian replied:
I didn't see a test for those changes, did I miss it?

kmaglione replied:
On Tue, May 05, 2015 at 08:59:39AM -0700, Mathieu Agopian wrote:
>>
>>
>>  def build_quick_xpcom(method, interface, traverser, wrapper=False):
>>      """A shortcut to quickly build XPCOM objects on the fly."""
>> +    extra = ()
>
>Seems like it got quite a few changes, can you please document what it's doing now? Constructing "deep" objects?

Essentially, I need these objects to expose multiple interfaces. 
There's support elsewhere for doing this when JavaScript 
requests them, e.g.,

```js
prefService.QueryInterface(Ci.nsIPrefBranch).QueryInterface(Ci.nsIPrefService);
```

But no way from doing so from Python that isn't a chore. I made 
several attempts to do this cleanly and without any duplication, 
but too much of the related code is such a mess that I didn't 
have the time. The logic in the `for iface in extra` loop is 
copied and simplified from `xpcom_constructor`.

I started working on another PR for jQuery validation a while 
ago that should move some of that logic to somewhere more 
accessible, and in the end simplify both of these functions.

kmaglione replied:
On Tue, May 05, 2015 at 09:00:04AM -0700, Mathieu Agopian wrote:
>>
>>
>>  def build_quick_xpcom(method, interface, traverser, wrapper=False):
>>      """A shortcut to quickly build XPCOM objects on the fly."""
>> +    extra = ()
>
>I didn't see a test for those changes, did I miss it?

There really aren't any tests for most of the validator 
internals. I'm not even sure where I'd put it. I did add tests 
for the services in `Services` object (for which I implemented 
this), which check that the behavior is the same whether the 
services is instantiated in JS (as above), or accessed via the 
services object, so this is being tested indirectly.

mstriemer commented on validator/testcases/javascript/predefinedentities.py:
```diff
@@ -121,8 +216,18 @@
-            {u"currentEngine": {"readonly": True},
-             u"defaultEngine": {"readonly": True}}},
+            {u"currentEngine":
+                {"readonly": search_warning(severity="high")},
```
Shouldn't `readonly` be `True` or `False`? Should the keys here be `"dangerous"` instead?

kmaglione replied:
`"dangerous"` applies to any access, or in the case of functions, to calls. `"readonly"` only applies to attempts to write to the value, which is what's important for these cases.

As for `True` or `False`, `"readonly"` has accepted `True`/`False`/`lambda` for a long time. This patch changes it to also accept a message description or a `dict` of message properties the same way `"dangerous"` does.

kmaglione commented:
Pushed updates in a second commit, in case that's easier to review. Will rebase into a single commit before merge, or sooner if you'd prefer.

This now has nearly all of the planned tests. There are a few which I haven't come up with a satisfactory way of testing (mainly because they're highly context-sensitive) and am considering pushing back until after the initial roll-out, or entirely into runtime testing.

magopian commented:
Huge work, impressive. What's missing?

I see there's a couple of `FIXME`s, should they be fixed before merging?

magopian commented on validator/testcases/regex.py:
```diff
@@ -2732,3 +2736,113 @@ def tests(self):
+    (DANGEROUS_CATEGORIES, DANGEROUS_CATEGORY_WARNING),
+)
+
+validate_string = RegexTest(STRING_REGEXPS).test
```
Given the complexity of the code, and the easiness of making mistakes with regular expressions, I think those should be tested (talking about munge_filename, PROFILE_REGEX, validate_string, and RegextTest).

kmaglione replied:
On Mon, May 18, 2015 at 07:03:13AM -0700, Mathieu Agopian wrote:
>Given the complexity of the code, and the easiness of making mistakes with regular expressions, I think those should be tested (talking about munge_filename, PROFILE_REGEX, validate_string, and RegextTest).

You're probably right. I added tests for all of the permutations 
that these tests should match, but unit tests would definitely 
not hurt. Will do.

kmaglione commented:
On Mon, May 18, 2015 at 06:33:43AM -0700, Mathieu Agopian wrote:
>Huge work, impressive. What's missing?



>I see there's a couple of `FIXME`s, should they be fixed before merging?

I think all of the FIXMEs I added are relevant to issues in old 
code that I happened to stumble across, but not particularly 
relevant to these changes. Adding support to treat `instanceof` 
like `QueryInterface` would definitely help some of these tests, 
though.

-- 
Kris Maglione

For every complex problem there is an answer that is clear, simple,
and wrong.
	--H. L. Mencken

kmaglione commented:
On Mon, May 18, 2015 at 06:33:43AM -0700, Mathieu Agopian wrote:
>What's missing?

From my spec:

• Changes to #identity-box, #searchbar, or #search-container

Of these, I could check for references to #identity-box, but I 
can't check for sub-elements as easily, and I haven't found any 
example code making these kinds of changes, so I'm not entirely 
sure any checks I come up with would be worth it.

#search-container is, unfortunately, referenced so often, and in 
so may different ways, by add-ons trying to insert things next 
to it, that I haven't been able to think of any non-runtime 
checks that would be useful. #searchbar has similar problems to 
#identity-box. I know there are add-ons that modify it, but I 
don't have access to code that would give me a good idea of how.

• Add-on uninstall detection.

The API doesn't make this especially easy. The method is just 
named `uninstall`, and trying to figure out what objects may be 
add-on objects is just too much work without massive changes to 
the validator. In any case, this is more of a nice-to-have. 
Add-on installation is really the important case.

• `importScripts` with remote scripts.

I'm considering adding a check for static strings that are 
remote, but I mainly have the impression that this isn't used 
very often with static strings.

• `evalInSandbox` on sandboxes created with chrome principals.

After giving this some more thought, I think I'm going to add a 
check for any `evalInSandbox` usage. But trying to figure out if 
the sandbox was chrome-privileged turned out to be far too 
difficult.

• href="javascript:..."

Same as above, I'd mainly intended this to apply to chrome 
scopes, but I think I'll add some more general checks.

• Accessing anything outside of ProfD/TempD (how do we do this statically?)

I haven't yet come up with an answer to "how do we do this 
statically", aside from prohibiting `nsIFile` objects from being 
created with `initWithPath` when unnecessary, and then 
restricting what paths can be gotten from the directory service. 
But there are still other file access APIs, and implementing all 
of the necessary semantics would be extremely complicated.

-- 
Kris Maglione

I live in the world rather as a spectator of mankind than as one of
the species.
	--Joseph Addison

magopian commented:
So what are the next steps? Fixing the tests, and then releasing the current work?

kmaglione commented:
If you don't have any other comments, I'll write tests for the new regex test class and the checks I mentioned above (which should be trivial), and then I think we're good.

I think fixing the failing tests just involves rebasing onto master. They pass locally. All the failures here are from a missing spidermonkey executable.

kmaglione commented:
Alright, Travis tests fixed, commits squashed. Should be good to go.

magopian commented:
Thanks a lot for this huge work @kmaglione!
